content/blog/client-futures.md

```diff
@@ -0,0 +1,72 @@
+
+If you define your own futures, then you need to update their definitions too, considering whether you need to handle errors.
+
+There is a `TryFuture` type in futures 0.3 which is roughly equivalent to `Future<Output=Result<...>>`. However, it is best to avoid it as much as possible - using it means you have to convert between `Future` and `TryFuture`. It has a blanket impl, which makes it easy to use some functions on these kind of futures by using the `TryFutureEx` trait.
+
+`Future::poll` takes a new context argument in futures 0.3. Mostly, this just needs propagating through `poll` calls (and occasionally ignoring).
+
+Our dependencies are still using futures 0.1, so we have to convert between the two libraries. This is mostly not too difficult because futures 0.3 includes some compatibility shims and other utilities (`Compat01As03`, etc.). These must be used where we call into our dependencies.
+
+The `wait` method has been removed from the `Future` trait, this was a good idea since it is a bit of a footgun. It can be replaced by either `.await` or `executor::block_on` (note that the latter blocks the whole thread, not just the current future).
+
+
+## Pin
+
+Futures 0.3 makes heavy use of the [`Pin`](https://doc.rust-lang.org/nightly/std/pin/index.html) type, in particular for the type of `self` in the signature of `Future::poll`. Other than the mechanical change to signatures, I had to do a bit of juggling using `Pin::get_unchecked_mut` and `Pin::new_unchecked` (both unsafe) to project fields of futures.
+
+Pinning is a subtle and complicated concept. I still don't feel like I have a great grasp of it all. The best reference I'm aware of are the [std::pin docs](https://doc.rust-lang.org/nightly/std/pin/index.html). I'll try and summarise the ideas here (there are some important subtleties that I'm not going to touch on at all, this is not intended to be a tutorial on pinning):
+
+* `Pin` is a type constructor which only makes sense when applied to a pointer type, e.g., `Pin<Box<_>>`.
+* `Pin` itself is a marker/wrapper type (a bit like [`NonNull`](https://doc.rust-lang.org/nightly/std/ptr/struct.NonNull.html)), not a pointer type. You should read `Pin<Box<Foo>>` as 'a pinned Box pointer to Foo' or 'a Box<Foo> which is guaranteed to be pinned'.
+* If a pointer is pinned, that means that the value pointed to by the pointer will never be moved (moving happens when a non-`Copy` object is passed by value, or via `mem::swap`). Note that the value is allowed to move *before* the pointer is pinned, but not *after*.
```

Comment:
```suggestion
* If a pointer is pinned, it means that the value pointed to by the pointer will never be moved (moving happens when a non-`Copy` object is passed by value, or via `mem::swap`). Note that the value is allowed to be moved *before* the pointer is pinned, but not *after*.
```

"by value" is a term of art which means a value is copied, rather than a reference is copied (which is called "by reference").

Reply:
Good to know.